### DISCUSSION TITLE
quarkus azure functions

### DISCUSSION THREAD

**Discussion by petar1201** [quarkusio/quarkus#32820] [Community]
How does exactly new extension work, since i am still not able to work with Service Bus triggers in 3.0.0.CR2 , is full functionality expected in 3.0.0.Final version and then we can use Service bus triggers in quarkus?

**Comment by maxandersen** [comment]
Wdym new extension? Something specific you are referring to?

**Reply by petar1201** [reply]
Maybe i am not naming it correctly, but look at #30762 , they are reffering to it as azure functions extension. Anyways, i mean about new library that was introduced to us in quarkus 3 relaeases, as i understand, before, we were able to use quarkus only with http trigger, but now with new quarkus-azure-functions library, as i read, we should be able to use quarkus with another types of triggers, for example i need it to work with service bus topic trigger. So in my question i am reffering to usage of that specific library and speaking of a problem not being able to work with different types of triggers, current version of quarkus available and of quarkus-azure-functions is 3.0.0.CR2, and as i said it doesn't seem to work for me yet. For example, quarkus-azure-functions-http, that was used earlied, should be used in a specific way, where each function's entry point is 
"io.quarkus.azure.functions.resteasy.runtime.Function.run", so than it does it's logic. Since we don't have much of docmentation, i would really like to know how does exactly quarkus-azure-functions work and how should it be used, since only example we can find is again with http trigger, is there any specific stuff we should do, or should new library do it's logic in the background? Sorry for over writing, but i tried to explain core of a problem the best as i can!

**Reply by maxandersen** [reply]
Cc @patriot1burke - he should be able to help.

**Comment by patriot1burke** [comment]
quarkus-azure-functions-http extension is a bridge to use Resteasy, Vert.x Web, Servlet, or our Spring MVC layer with azure functions.

The quarkus-azure-functions extension allows you to use quarkus and azure functions directly with any azure function trigger type.  The function class will be a CDI bean and you can use things like @Inject.  Quarkus will also boot automatically.

Does that answer your question? 

You'd have to explain more why you think the quarkus-azure-functions extension doesn't work with service bus.  Are you using @ServiceBusQueueTrigger or TopicTrigger?

**Reply by petar1201** [reply]
I am using @ServiceBusTopicTrigger, and i will describe you particular behaviour i am receiving. I am using kotlin with gradle, and when i dont't add dependency for quarkus-azure-functions:3.0.0.CR2, and send message to topic, it gives some info about trigger, does the function, and than says that it is executed, and for each message i send, these three steps are done corectly. When i add quarkus-azure-functions:3.0.0.CR2 dependency to my build.gradle.kts file, without changing anyhing in code, now when i send message to topic, it only does the first step correctly, meaning, it gives me info about trigger, but it doesn't proceed to actual function, and that happens for each message i send. I thought maybe adding **@ApplicationScoped** annotation can change something since Function class will be CDI bean but that didn't change anything. This kind of behaviour happens to me even with http trigger when adding quarkus-azure-functions extension to project, i didn't deploy anything and ran only localy. Maybe i am doing something else wrong since even http trigger doesn't work, but i really don't see where could mistake be, since i only have basic properties set, and as i said, without this library, triggers actualy work, but then obviusly quarkus isnt't bootstraped

**Comment by petar1201** [comment]
Hello @patriot1burke, @maxandersen, i solved my problem and now everything is working properly, making the fat jar solved the problem. If you want i can create a document with example and readme with details, how it works and how it should be done, so when others try to do this, they do not experience same problems i did?

**Comment by patriot1burke** [comment]
Hmmm, the extension is supposed to enforce a legacy thin-jar deployment which should work.

**Comment by patriot1burke** [comment]
I'd rather fix the extension to package things up properly.

**Comment by petar1201** [comment]
yeah, that is probably better and nicer way. But i don't think that problem is in quarkus-azure-functions extensions, maybe in some layer when building, i don't know how quarkus building tasks are made(if i am naming it corectly), but maybe something can be changed there. Or you can try and test again, i said that in my project even http trigger didn't work, and you gave a working example of http trigger, so maybe it is about some specific property or something that changes things. As i said in some of my previous comments, documentation is not on level it should be, so if there is something that needs to be set or works in a specific way, you can update and give better docs, but if you test it and it doesn't work, than problem is problably with how azure works and what kind of jar needs to be. Hope you understood what i wanted to say, as i said, i can help you in terms of giving my example and problems i enroled while trying every single thing and property and this finaly did the trick.

**Comment by patriot1burke** [comment]
Could you post the pom.xml you had problems building/deploying with before you force a fat jar?

**Reply by petar1201** [reply]
I am using gradle, so i can post you my build.gradle.kts and my gradle.properties.

This is version before doing it with fat/uber jar. application.properties file is empty in this version.
Old version:
[oldVersion.zip](https://github.com/quarkusio/quarkus/files/11311215/oldVersion.zip)

In the version that is working, gradle.properties file didn't change, but something is added in build.gradle.kts, and in application.properties.
New(working version):
[newVersion.zip](https://github.com/quarkusio/quarkus/files/11311240/newVersion.zip)

**Comment by petar1201** [comment]
Hello @patriot1burke , @maxandersen , i want to inform you that problem wasn't because of fat jar, what actually made the difference is -runner sufix that is created automatically on jar file in Quarkus 2.

I have been testing out Service Bus triggers, and HTTP triggers with Quarkus 2 and Quarkus 3 releases. Locally there weren't any problems with both versions, but when i deployed my function to Azure (i am using Azure Pipelines), with Quarkus 2, everything is ok, but when i used Quarkus 3, i couldn't actually trigger (execute) the function, since for every trigger i made, i kept getting ClassNotFoundException. Note that only thing that was changed is only the version of Quarkus, nothing else in code and project itself. Deploy is indeed successful, and i can see in portal both my function and its associated files, but when it comes to execution of Function it fails.

Have you tried deployment of Azure Functions with Quarkus 3 yourselves on Azure Portal? Do you maybe have any idea or solution for this?

Hope to hear from you soon!
